### ISSUE TITLE
/etc/apk/repositories defaults to https

### ISSUE THREAD

**Issue by fiesh** [alpinelinux/docker-alpine#184]
From `3.12` to `3.13`, the default in `/etc/apk/repositories` was changed from `https` to `http`.

Despite people thinking this is the way to go (see #139), this is an anti-feature in that it
* provides no additional security at all,
* introduces overhead,
* defies caching proxies used in CI pipelines.

Other distributions don't do that, so the argument that corporate policies might require this seems to be void at least of now.  There is however a somewhat ill-devised _move everything to https_ trend happening, github being a prime example where downloads only work via `https` now.

I'd like to ask to revert this change and possibly clarify in the documentation why `http` is the correct protocol to use here.

**Comment by sebma** [comment]
@fiesh I agree with [ncopa's comment](https://github.com/gliderlabs/docker-alpine/issues/184#issuecomment-874033622) that says : 

> The main reason for moving to https was that it "solved" some network issues for some users, which I suspect was related to intercepting proxies.

I had this pb. myself, I [couldn't install anything with apk/HTTP](https://github.com/alpinelinux/docker-alpine/issues/132) because of the proxy of the school.

Does HTTPS keep you from installing apk packages through the network ?

**Comment by fiesh** [comment]
Thanks for your reply, @sebma 

It does not keep me from using it.  It does however severely impair our CI pipelines since it prohibits the usage of our caching proxy.

While I understand that there are plenty of ill configured networks that might lead to trouble, this seems like a workaround for broken firewall choices that hurt users in a normal environment.  It would seem more natural to make those in a broken environment having to work around their issues with a `sed -i -e 's/http/https/g' /etc/apk/repositories` to me.

**Comment by AndreKoepke** [comment]
From your points on https://github.com/gliderlabs/docker-alpine/issues/184 - I don't want to open a big discussion, but would like to say, that HTTPS offers integrity. [Ask IBM.](https://www.ibm.com/docs/en/ibm-mq/9.1?topic=tls-how-provides-identification-authentication-confidentiality-integrity)

@fiesh maybe you can contact me private (e.g. per mail, akop_github[at]akop.online), so we can discuss it directly. :)

I vote to close this ticket, because some the preconditions from fiesh are incorrect.  
The reasons:
 * Additional security, because of all points TLS (encryption, integration, trusted servers, ...)
 * Yes, there is more CPU-time used. But it isn't much
 * Caching proxies can be used (via normal web-proxies)
 * My company also enforcing HTTPS, the "work around" isn't nice when pretty much people have to applied it.

@fiesh HTTPS is the future. I think, your caching proxy is out of date, when it can't handle it.

**Comment by fiesh** [comment]
@AndreKoepke You are misunderstanding integrity here.  TLS does provide integrity for the transmission of data.  It does not provide integrity for the underlying data itself.

As for your comment, there is no proxy that can cache `https`.  That's the very design goal of `https`.  If there was, `https` would be flawed.  You may think it's the future, but you stress the misconceptions that people have with regards to `https` and encryption in general.

**Comment by sebma** [comment]
@fiesh As @AndreKoepke suggested, you should discuss those proxy details  privately.

And if you really do want HTTP, maybe you can open a new issue as a feature request or a PR proposing a new `apk` option such as `apk --no-ssl-tls`.

**Comment by fiesh** [comment]
@sebma I mentioned proxying as a legitimate advantage of `http` over `https` -- how is this not a meaningful argument?

It seems to me the positions are:

* `https` benefits restrictive network configurations
* `http` benefits open network configurations

I'm not sure how preferring the former is a good design choice.  But if that's the consensus here, then we can close the issue obviously.

**Comment by fiesh** [comment]
(Adding this option doesn't seem like a good decision to me since then `apk` somehow needs to determine different sources based on command line options.  This seems complicated and fragile to me.)

**Comment by sebma** [comment]
@fiesh 
> I mentioned proxying as a legitimate advantage of http over https -- how is this not a meaningful argument?

It is but using HTTPS was just a workaround for me to apk work through the school proxy.

**Comment by fiesh** [comment]
> It is but using HTTPS was just a workaround for me to apk work through the school proxy.

I understand that and see the argument, but like I said a workaround shouldn't be imposed on all users I think, in particular if it brings a disadvantage to them.

**Comment by sebma** [comment]
@fiesh 

> (Adding this option doesn't seem like a good decision to me since then `apk` somehow needs to determine different sources based on command line options. This seems complicated and fragile to me.)

Less complicated that having the network admin able to configure his proxy correctly to be able to use apk through a proxy in my opinion.

**Comment by fiesh** [comment]
@sebma

> Less complicated that having the network admin able to configure his proxy correctly to be able to use apk through a proxy in my opinion.

I'm not sure what that means.  Using `http` doesn't entail any requirements on proxying or whatever, it simply allows for it if wanted.  Thus it's the better default in my opinion.

**Comment by sebma** [comment]
@fiesh I agree. Maybe HTTP can be the default.

How about we ask for a new `apk --ssl` or `apk --tls` option for the people who need HTTPS for apk to work through school/university misconfigured proxies ?

**Comment by fiesh** [comment]
@sebma 

> @fiesh I agree. Maybe HTTP can be the default.
> 
> How about we ask for a new `apk --ssl` or `apk --tls` option for the people who DO need HTTPS for apk to work through school/university misconfigured proxies ?

I'm all for `http` being the default but have my doubts such an option will make it ;)  But I'm happy to participate in the discussion if you want to open up an issue for it!

**Comment by AndreKoepke** [comment]
I was wrong. HTTPS allows proxing, but the proxy can't read the transmitted data. That's why I love discussions - break the own filter bubble. :)

But unencrypted traffic should be still avoided. Maybe a `--mirror http[s]://my-apt-mirror` flag like docker/maven/gradle is the best option. So we're getting the best out of all worlds.

**Comment by fiesh** [comment]
@sebma 

Maybe a filtering option for the transport layer would be a good command line option.  Then `/etc/apk/repositories` could contain different sources like `http`, `https`, and lets say `ftp` for the sake of the argument, and potentially others in the future.  And `apk --allow-transport https` or so could be used to only allow `https` transports.

**Comment by fiesh** [comment]
@AndreKoepke _But unencrypted traffic should be still avoided._  This is a personal preference of yours that I do not share and for which I see no technical argument.  Thus it should be stated as your view, not an absolute truth ;)

**Comment by AndreKoepke** [comment]
Yes, it is my personal opinion. The technical arguments are still [here](https://github.com/alpinelinux/docker-alpine/issues/184#issuecomment-874066363).

The signing-mechanism of `apk` should be fine for the official repos, but was is with the custom repos?  
And one locked door (package signing) is fine, but a second door (TLS) is not a bad thing. It is still possible, that the signing-process has undiscovered exploits. But that the signing-process **and** TLS having exploits is very unlikely.

TLS also offers protection for reading traffic. Maybe a MITM-attacker found out that you use a outdated package with open CVEs. Yes, it sounds a little bit crazy, but it is still a vector. ^^

**Comment by sebma** [comment]
> I don't think it's worth switching back to http. My impression is that the number of users who have problems with http is bigger than the number of users having problems with https, and it is relatively simple for those who needs http to sed /etc/apk/repositories.

I'm so glad to read that because I, also, had problems using `apk` over HTTP through the schools' proxy.

**Comment by jameyg42** [comment]
i realize this issue is closed, but the issue i've run into is when https:// is enabled, but the ca-certificates package isn't available.  In order to use https://, i first need to add the cert for the corporate proxy, but there's no clean way to do that if ca-certificates isn't available.

**Comment by pfeigl** [comment]
We are facing the very same problem @jameyg42 described. 

With `ca-certificates` not beeing part of the base image, it's practically impossible to install corporate certificates in a controlled way, where the certificates are already needed to talk to the proxy. 
Ofcourse you can `>> /etc/ssl/certs/ca-certificates.crt`, but that's not really desirable when there is something available like `update-ca-certificates`.

So my vote would be so simply put ca-certificates` package into the base alpine images to enable those usecases.
